File: oop/06-inheritance-vs-composition/csharp/Inheritance/InterestBearingAccount.cs
/*
 * Demonstrates: Inheritance vs Composition - Inheritance Approach
 * Concept: Base class that provides interest calculation behaviour through inheritance
 */

namespace FinanceDomain.InheritanceVsComposition.Inheritance;

public abstract class InterestBearingAccount
{
    public string AccountNumber { get; }
    public string AccountHolder { get; }
    protected decimal _balance;
    public decimal InterestRate { get; }

    public decimal Balance => _balance;

    protected InterestBearingAccount(string accountNumber, string accountHolder, decimal openingBalance, decimal interestRate)
    {
        AccountNumber = accountNumber;
        AccountHolder = accountHolder;
        _balance = openingBalance;
        InterestRate = interestRate;
    }

    public void Deposit(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        _balance += amount;
    }

    public virtual void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        if (amount > _balance)
            throw new InvalidOperationException($"Insufficient funds. Available: R{_balance:F2}");

        _balance -= amount;
    }

    public decimal CalculateInterest()
    {
        return _balance * InterestRate;
    }

    public void ApplyInterest()
    {
        decimal interest = CalculateInterest();
        _balance += interest;
    }
}

File: oop/06-inheritance-vs-composition/typescript/inheritance/interestBearingAccount.ts
/*
 * Demonstrates: Inheritance vs Composition - Inheritance Approach
 * Concept: Base class that provides interest calculation behaviour through inheritance
 */

export abstract class InterestBearingAccount {
    public readonly accountNumber: string;
    public readonly accountHolder: string;
    protected _balance: number;
    public readonly interestRate: number;

    constructor(accountNumber: string, accountHolder: string, openingBalance: number, interestRate: number) {
        this.accountNumber = accountNumber;
        this.accountHolder = accountHolder;
        this._balance = openingBalance;
        this.interestRate = interestRate;
    }

    get balance(): number {
        return this._balance;
    }

    deposit(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }
        this._balance += amount;
    }

    withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }
        if (amount > this._balance) {
            throw new Error(`Insufficient funds. Available: R${this._balance.toFixed(2)}`);
        }
        this._balance -= amount;
    }

    calculateInterest(): number {
        return this._balance * this.interestRate;
    }

    applyInterest(): void {
        const interest = this.calculateInterest();
        this._balance += interest;
    }
}
